browser.test_cross_origin regression: "Expected to find '/report_result?exit:0 ' in '/report_result?abort:Assertion failed"

After PR #27018

```
commit 41656d690b521208189ed838c5ab474435b208c9
Author: Sam Clegg <sbc@chromium.org>
Date:   Thu May 28 14:29:24 2026 -0700

    [pthreads] Simplify message relaying code (#27018)

    Historically, `targetThread` relaying was used to pass `objectTransfer`
    messages (containing `OffscreenCanvas` objects) back to parent threads
    on worker exit. This was introduced in commit 1b278917cd (2018/2019) and
    required supporting `transferList` in the relaying code.

    However, automatic canvas return on exit was later disabled in
    6dcf7298e8 (2019) and the `objectTransfer` command was fully removed in
    0f536c60bb (2021).

    These days, the relaying mechanism is only used by `CMD_CHECK_MAILBOX`
    (via `_emscripten_notify_mailbox_postmessage`) to send simple mailbox
    notifications, which do not require a transfer list.
```

the test `browser.test_cross_origin` (and possibly several other tests, as the Firefox suites are now very red) are failing with

```

C:\emsdk\emscripten\main>test\runner browser.test_cross_origin
Running 1 tests
[1/1] test_cross_origin (test_browser.browser.test_cross_origin) ... FAIL

Stderr:
127.0.0.1 - - [31/May/2026 20:27:49] code 404, message File not found
127.0.0.1 - - [31/May/2026 20:27:49] "GET /hello.js HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2026 20:27:49] "GET /hello.wasm HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2026 20:27:49] "GET /hello.js HTTP/1.1" 200 -


======================================================================
FAIL: test_cross_origin (test_browser.browser.test_cross_origin)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\emsdk\emscripten\main\test\common.py", line 380, in resulting_test
    return func(self, *args)
  File "C:\emsdk\emscripten\main\test\test_browser.py", line 5659, in test_cross_origin
    self.run_browser('test.html', '/report_result?exit:0')
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\emsdk\emscripten\main\test\browser_common.py", line 899, in run_browser
    raise e
  File "C:\emsdk\emscripten\main\test\browser_common.py", line 890, in run_browser
    self.assertContained(expected, output)
    ~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^
  File "C:\emsdk\emscripten\main\test\common.py", line 1136, in assertContained
    self.fail("Expected to find '%s' in '%s', diff:\n\n%s\n%s" % (
    ~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      limit_size(values[0]), limit_size(string), limit_size(diff),
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      additional_info,
      ^^^^^^^^^^^^^^^^
    ))
    ^^
AssertionError: Expected to find '/report_result?exit:0
' in '/report_result?abort:Assertion failed
', diff:

--- expected
+++ actual
@@ -1 +1 @@
-/report_result?exit:0
+/report_result?abort:Assertion failed



Stderr:
127.0.0.1 - - [31/May/2026 20:27:49] code 404, message File not found
127.0.0.1 - - [31/May/2026 20:27:49] "GET /hello.js HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2026 20:27:49] "GET /hello.wasm HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2026 20:27:49] "GET /hello.js HTTP/1.1" 200 -

----------------------------------------------------------------------
Ran 1 test in 2.475s

FAILED (failures=1)

C:\emsdk\emscripten\main>
```

--- a/src/lib/libpthread.js
+++ b/src/lib/libpthread.js
@@ -286,11 +286,7 @@ var LibraryPThread = {
         // If this message is intended to a recipient that is not the main
         // thread, forward it to the target thread. This is currently only
         // used by `CMD_CHECK_MAILBOX`.
-        if (d.targetThread) {
-#if ASSERTIONS
-          // pthreads should not be relaying messages to themselves.
-          assert(d.targetThread != _pthread_self());
-#endif
+        if (d.targetThread && d.targetThread != _pthread_self()) {
           var targetWorker = PThread.pthreads[d.targetThread];
 #if ASSERTIONS
           if (!targetWorker) err(`worker sent message (${cmd}) to pthread (${d.targetThread}) that no longer exists`);

--- a/test/codesize/test_codesize_minimal_pthreads.json
+++ b/test/codesize/test_codesize_minimal_pthreads.json
@@ -1,10 +1,10 @@
 {
-  "a.out.js": 6945,
-  "a.out.js.gz": 3424,
+  "a.out.js": 6955,
+  "a.out.js.gz": 3431,
   "a.out.nodebug.wasm": 19108,
   "a.out.nodebug.wasm.gz": 8824,
-  "total": 26053,
-  "total_gz": 12248,
+  "total": 26063,
+  "total_gz": 12255,
   "sent": [
     "a (memory)",
     "b (exit)",

--- a/test/codesize/test_codesize_minimal_pthreads_memgrowth.json
+++ b/test/codesize/test_codesize_minimal_pthreads_memgrowth.json
@@ -1,10 +1,10 @@
 {
-  "a.out.js": 7379,
-  "a.out.js.gz": 3639,
+  "a.out.js": 7389,
+  "a.out.js.gz": 3647,
   "a.out.nodebug.wasm": 19109,
   "a.out.nodebug.wasm.gz": 8826,
-  "total": 26488,
-  "total_gz": 12465,
+  "total": 26498,
+  "total_gz": 12473,
   "sent": [
     "a (memory)",
     "b (exit)",

--- a/test/test_browser.py
+++ b/test/test_browser.py
@@ -3954,6 +3954,12 @@ Module["preRun"] = () => {
   def test_pthread_run_on_main_thread(self):
     self.btest_exit('pthread/test_pthread_run_on_main_thread.c', cflags=['-O3', '-pthread', '-sPTHREAD_POOL_SIZE'])
 
+  # Test that proxying operations work when Atomics.waitAsync is disabled,
+  # forcing the waitAsyncPolyfilled/postMessage fallback path.
+  def test_pthread_no_waitasync(self):
+    create_file('pre.js', 'delete Atomics.waitAsync;\n')
+    self.btest_exit('pthread/test_pthread_proxy_to_pthread.c', cflags=['-O3', '-pthread', '-sPROXY_TO_PTHREAD', '-sASSERTIONS', '--pre-js=pre.js'])
+
   # Test how a lot of back-to-back called proxying operations behave.
   def test_pthread_run_on_main_thread_flood(self):
     self.btest_exit('pthread/test_pthread_run_on_main_thread_flood.c', cflags=['-O3', '-pthread', '-sPTHREAD_POOL_SIZE'])

